File: atcoder_selected_100/ruby/s8pc_1_g.rb
# 50
# S8PC-1/G
# 動的計画法:bit DP
# 巡回セールスマン問題

# 最短時間 == 最短距離
N, M = gets.split.map(&:to_i)
D = Array.new(N) { Array.new(N, Float::INFINITY) }
T = Array.new(N) { Array.new(N, 0) }

M.times do
  s, t, d, time = gets.split.map(&:to_i)
  s -= 1
  t -= 1
  D[s][t] = D[t][s] = d
  T[s][t] = T[t][s] = time
end

# dists[移動した建物のBIT][現在の建物]
dists = Array.new(1 << N) { Array.new(N, Float::INFINITY) }
cnts = Array.new(1 << N) { Array.new(N, 0) }
# 0スタートで0に戻ってくることからbit=0, i=0を初期化
dists[0][0] = 0
cnts[0][0] = 1

(1 << N).times do |bit|
  N.times do |i|
    next if bit[i] == 1 # すでにjを通っている場合
    N.times do |j|
      next if i == j
      dist = dists[bit][j] + D[j][i]
      next if T[j][i] < dist # 通れる時間を過ぎている場合
      next_bit = bit | (1 << i)
      case dists[next_bit][i] <=> dist
      when 1
        dists[next_bit][i] = dist
        cnts[next_bit][i] = cnts[bit][j]
      when 0
        cnts[next_bit][i] += cnts[bit][j]
      else
        nil
      end
    end
  end
end

puts cnts[-1][0].zero? ? "IMPOSSIBLE" : [dists[-1][0], cnts[-1][0]].join(" ")

File: atcoder_selected_100/kotlin/s8pc_1_g.kt
/**
 * 50
 * S8PC-1/G
 * 動的計画法:bit DP
 * 巡回セールスマン問題
 */

fun main() {
  val (N, M) = readLine()!!.split(" ").map(String::toInt)
  val G = Array(N) { mutableListOf<Triple<Int, Long, Long>>() }
  repeat(M) {
    val (s, t, d, time) = readLine()!!.split(" ")
    G[s.toInt() - 1].add(Triple(t.toInt() - 1, d.toLong(), time.toLong()))
    G[t.toInt() - 1].add(Triple(s.toInt() - 1, d.toLong(), time.toLong()))
  }

  // dp[現在の建物][通った建物のbit] = Pair<距離, 総数>
  val dp = Array(N) { Array(1.shl(N)) { Pair(Long.MAX_VALUE, 0L) } }
  dp[0][0] = Pair(0L, 1L)

  repeat(1.shl(N)) { bit ->
    repeat(N) i@ { i ->
      if (dp[i][bit].first == Long.MAX_VALUE) return@i
      G[i].forEach { (j, d, time) ->
        if (bit.shr(j).and(1) == 1) return@forEach
        val dist = dp[i][bit].first + d
        if (dist > time) return@forEach
        val next_bit = bit + 1.shl(j)
        when (dp[j][next_bit].first.compareTo(dist)) {
          1 -> dp[j][next_bit] = Pair(dist, dp[i][bit].second)
          0 -> dp[j][next_bit] = Pair(dp[j][next_bit].first, dp[j][next_bit].second + dp[i][bit].second)
        }
      }
    }
  }
  val bit_max = 1.shl(N) - 1
  println(if (dp[0][bit_max].second == 0L) "IMPOSSIBLE" else dp[0][bit_max].toList().joinToString(" "))
}
